File: Visual/XSharp/Core/CorePrep/Verifier.cpp
// SPDX-FileCopyrightText: 2026 Leitwolf <support@xsharp-lang.xyz>
// SPDX-License-Identifier: MPL-2.0

#include "Visual/XSharp/Core/CorePrep/Verifier.hpp"
#include "Visual/XSharp/Core/CorePrep/Verifier/Semantics.hpp"

#include <algorithm>
#include <unordered_set>

namespace visual_xsharp::core
{
namespace
{
auto issue(std::string code, std::string message, SymbolId function, BlockId block = 0) -> VerificationIssue
{
    return VerificationIssue{std::move(code), std::move(message), function, block};
}

auto atom_valid(const Atom &atom) -> bool
{
    if(atom.kind == Atom::Kind::Variable)
        return atom.symbol.id != 0;
    switch(atom.type.kind)
    {
    case Type::Kind::Unit: return std::holds_alternative<std::monostate>(atom.literal);
    case Type::Kind::Bool: return std::holds_alternative<bool>(atom.literal);
    case Type::Kind::Int64: return std::holds_alternative<std::int64_t>(atom.literal);
    case Type::Kind::Int32: return std::holds_alternative<std::int32_t>(atom.literal);
    case Type::Kind::String: return std::holds_alternative<std::u32string>(atom.literal);
    case Type::Kind::Function:
    case Type::Kind::Named:
    case Type::Kind::TypeVariable: return false;
    }
    return false;
}

auto expected_arity(Operation operation) -> std::size_t
{
    switch(operation)
    {
    case Operation::Copy: return 1;
    case Operation::Call: return 0;
    case Operation::Negate:
    case Operation::LogicalNot: return 1;
    default: return 2;
    }
}

void verify_instruction(const Instruction &instruction, SymbolId function, BlockId block,
                        std::vector<VerificationIssue> &issues)
{
    if(instruction.kind != Instruction::Kind::Evaluate && instruction.destination.id == 0)
        issues.push_back(issue("VXC1006", "defining instruction has no destination symbol", function, block));
    if(instruction.operation != Operation::Call && instruction.operands.size() != expected_arity(instruction.operation))
        issues.push_back(issue("VXC1007", "operation has the wrong operand count", function, block));
    if(instruction.operation == Operation::Call && instruction.operands.empty())
        issues.push_back(issue("VXC1008", "call operation has no callee operand", function, block));
    if(std::any_of(instruction.operands.begin(), instruction.operands.end(), [](const Atom &atom) { return !atom_valid(atom); }))
        issues.push_back(issue("VXC1009", "instruction contains an invalid typed atom", function, block));
}

void verify_terminator(const Terminator &terminator, const std::unordered_set<BlockId> &block_ids, SymbolId function,
                       BlockId block, std::vector<VerificationIssue> &issues)
{
    if(terminator.kind == Terminator::Kind::Return && !atom_valid(terminator.value))
        issues.push_back(issue("VXC1010", "return contains an invalid typed atom", function, block));
    if(terminator.kind == Terminator::Kind::Branch)
    {
        if(!atom_valid(terminator.value) || terminator.value.type.kind != Type::Kind::Bool)
            issues.push_back(issue("VXC1011", "branch condition must be a valid bool atom", function, block));
        if(!block_ids.contains(terminator.true_target) || !block_ids.contains(terminator.false_target))
            issues.push_back(issue("VXC1012", "branch targets a missing block", function, block));
    }
    if(terminator.kind == Terminator::Kind::Jump && !block_ids.contains(terminator.true_target))
        issues.push_back(issue("VXC1013", "jump targets a missing block", function, block));
}
} // namespace

auto verify(const CorePrepModule &module) -> std::vector<VerificationIssue>
{
    std::vector<VerificationIssue> issues;
    std::unordered_set<SymbolId> function_ids;
    for(const auto &function : module.functions)
    {
        if(function.symbol.id == 0 || !function_ids.insert(function.symbol.id).second)
            issues.push_back(issue("VXC1001", "function symbol is missing or duplicated", function.symbol.id));

        std::unordered_set<BlockId> block_ids;
        for(const auto &block : function.blocks)
            if(!block_ids.insert(block.id).second)
            issues.push_back(issue("VXC1002", "block id is duplicated", function.symbol.id, block.id));
        if(!block_ids.contains(function.entry))
            issues.push_back(issue("VXC1003", "entry block does not exist", function.symbol.id));

        std::unordered_set<SymbolId> definitions;
        for(const auto &parameter : function.parameters)
            if(parameter.symbol.id == 0 || !definitions.insert(parameter.symbol.id).second)
                issues.push_back(issue("VXC1004", "parameter symbol is missing or duplicated", function.symbol.id));
        for(const auto &block : function.blocks)
        {
            for(const auto &instruction : block.instructions)
            {
                if(instruction.kind == Instruction::Kind::Bind && !definitions.insert(instruction.destination.id).second)
                    issues.push_back(issue("VXC1005", "binding symbol is defined more than once", function.symbol.id, block.id));
                verify_instruction(instruction, function.symbol.id, block.id, issues);
            }
            verify_terminator(block.terminator, block_ids, function.symbol.id, block.id, issues);
        }
    }
    auto semantic_issues = verify_semantics(module);
    issues.insert(issues.end(), std::make_move_iterator(semantic_issues.begin()),
                  std::make_move_iterator(semantic_issues.end()));
    return issues;
}
} // namespace visual_xsharp::core

File: Compiler/Haskell/Core/src/Visual/XSharp/Core/CorePrep/Verifier.hs
-- SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
-- SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0
module Visual.XSharp.Core.CorePrep.Verifier (verifyCorePrep) where

import Data.List (nub)
import Visual.XSharp.AST
import Visual.XSharp.Core qualified as Core
import Visual.XSharp.Core.CorePrep
import Visual.XSharp.Diagnostic

verifyCorePrep :: CorePrepModule -> Either [Diagnostic] CorePrepModule
verifyCorePrep moduleValue =
    let problems =
            duplicateFunctions (corePrepModuleFunctions moduleValue)
                ++ concatMap verifyFunction (corePrepModuleFunctions moduleValue)
     in if null problems then Right moduleValue else Left problems

duplicateFunctions :: [CorePrepFunction] -> [Diagnostic]
duplicateFunctions functions = duplicateIds "VXC0001" "duplicate CorePrep function symbol" (map (resolvedSymbol . corePrepFunctionName) functions)

verifyFunction :: CorePrepFunction -> [Diagnostic]
verifyFunction function =
    let blocks = corePrepFunctionBlocks function
        blockIds = map corePrepBlockId blocks
        parameterIds = map (resolvedSymbol . fst) (corePrepFunctionParameters function)
        definitionIds = concatMap blockDefinitions blocks
     in missingEntry blockIds
            ++ duplicateIds "VXC0003" "duplicate CorePrep block id" blockIds
            ++ duplicateIds "VXC0004" "duplicate CorePrep parameter symbol" parameterIds
            ++ duplicateIds "VXC0005" "CorePrep symbol is defined more than once" (parameterIds ++ definitionIds)
            ++ concatMap (verifyBlock blockIds) blocks
    where
        missingEntry ids =
            if corePrepFunctionEntry function `elem` ids
                then []
                else [problem "VXC0002" "CorePrep function entry block does not exist"]

blockDefinitions :: CorePrepBlock -> [SymbolId]
blockDefinitions block = [resolvedSymbol name | CorePrepBind name _ _ _ <- corePrepBlockInstructions block]

verifyBlock :: [Int] -> CorePrepBlock -> [Diagnostic]
verifyBlock blockIds block =
    concatMap verifyInstruction (corePrepBlockInstructions block)
        ++ verifyTerminator blockIds (corePrepBlockTerminator block)

verifyInstruction :: CorePrepInstruction -> [Diagnostic]
verifyInstruction instruction = case instruction of
    CorePrepBind _ valueType _ operation -> verifyOperation valueType operation
    CorePrepAssign _ atom -> verifyAtom atom
    CorePrepEvaluate operation -> verifyOperation ErrorType operation

verifyOperation :: Type -> CorePrepOperation -> [Diagnostic]
verifyOperation resultType operation = case operation of
    CorePrepCopy atom -> verifyAtom atom ++ resultMismatch (atomType atom)
    CorePrepCall callee arguments -> verifyAtom callee ++ concatMap verifyAtom arguments
    CorePrepPrimitive primitive atoms -> concatMap verifyAtom atoms ++ verifyPrimitive primitive atoms resultType
    CorePrepMakeClosure function captures ->
        invalidFunction function
            ++ concatMap verifyCapture captures
            ++ closureResultMismatch resultType
    where
        resultMismatch valueType =
            if resultType == ErrorType || resultType == valueType
                then []
                else [problem "VXC0006" "CorePrep copy result type does not match its atom"]
        invalidFunction name = [problem "VXC0013" "CorePrep closure function symbol must be positive" | symbolIdValue (resolvedSymbol name) <= 0]
        closureResultMismatch valueType = case valueType of
            FunctionType _ _ -> []
            ErrorType -> []
            _ -> [problem "VXC0014" "CorePrep closure result must have a callable type"]

verifyCapture :: CorePrepCapture -> [Diagnostic]
verifyCapture (CorePrepCapture mode name valueType atom) =
    [problem "VXC0015" "CorePrep capture symbol must be positive" | symbolIdValue (resolvedSymbol name) <= 0]
        ++ [problem "VXC0016" "CorePrep capture has an unresolved type" | valueType == ErrorType]
        ++ [problem "VXC0017" "CorePrep capture atom type does not match its slot" | atomType atom /= valueType]
        ++ [ problem "VXC0018" "non-owning capture requires an AARC reference value"
           | mode /= StrongCapture && not (aarcReference valueType)
           ]
        ++ verifyAtom atom
    where
        -- Core currently lacks nominal-declaration metadata, so every resolved
        -- named type remains conservatively reference-like at this boundary.
        aarcReference value = case value of
            FunctionType _ _ -> True
            NamedType _ _ -> value /= unitType && value /= boolType && value /= intType
            _ -> False

verifyPrimitive :: Core.CorePrimitive -> [CorePrepAtom] -> Type -> [Diagnostic]
verifyPrimitive primitive atoms resultType
    | length atoms /= expectedArity = [problem "VXC0007" "CorePrep primitive has the wrong arity"]
    | any ((== ErrorType) . atomType) atoms = [problem "VXC0008" "CorePrep primitive contains an unresolved type"]
    | otherwise = operandProblems ++ resultProblems
    where
        expectedArity = if primitive `elem` [Core.CoreNegate, Core.CoreLogicalNot] then 1 else 2
        comparisonOrLogical =
            if primitive
                `elem` [ Core.CoreLessThan
                       , Core.CoreLessEqual
                       , Core.CoreGreaterThan
                       , Core.CoreGreaterEqual
                       , Core.CoreEqual
                       , Core.CoreNotEqual
                       , Core.CoreLogicalAnd
                       , Core.CoreLogicalOr
                       , Core.CoreLogicalNot
                       ]
                then True
                else False
        operandTypes = map atomType atoms
        firstType = case operandTypes of valueType : _ -> valueType; [] -> ErrorType
        operandsAgree = all (== firstType) operandTypes
        logical = primitive `elem` [Core.CoreLogicalAnd, Core.CoreLogicalOr, Core.CoreLogicalNot]
        numeric = all isNumericType operandTypes
        booleanContext = all (\valueType -> valueType == boolType || isNumericType valueType) operandTypes
        negatable = isSignedIntegerType firstType || isFloatingType firstType
        operandProblems
            | not operandsAgree = [problem "VXC0019" "CorePrep primitive operand types do not agree"]
            | logical && not booleanContext = [problem "VXC0020" "CorePrep logical primitive requires bool or numeric operands"]
            | primitive == Core.CoreNegate && not negatable =
                [problem "VXC0021" "CorePrep negation requires a signed integer or floating operand"]
            | not logical && not numeric && primitive `notElem` [Core.CoreEqual, Core.CoreNotEqual] =
                [problem "VXC0022" "CorePrep arithmetic or ordering primitive requires numeric operands"]
            | otherwise = []
        expectedResult = if comparisonOrLogical then boolType else firstType
        resultProblems = if resultType == expectedResult then [] else [problem "VXC0009" "CorePrep primitive result type is inconsistent"]

verifyTerminator :: [Int] -> CorePrepTerminator -> [Diagnostic]
verifyTerminator blockIds terminator = case terminator of
    CorePrepReturn atom -> verifyAtom atom
    CorePrepBranch atom trueTarget falseTarget -> verifyAtom atom ++ requireBool atom ++ targets [trueTarget, falseTarget]
    CorePrepJump target -> targets [target]
    CorePrepUnreachable -> []
    where
        targets values = [problem "VXC0010" "CorePrep terminator targets a missing block" | any (`notElem` blockIds) values]
        requireBool atom = [problem "VXC0011" "CorePrep branch condition must be bool" | atomType atom /= boolType]

verifyAtom :: CorePrepAtom -> [Diagnostic]
verifyAtom atom =
    [problem "VXC0012" "CorePrep atom has an unresolved type" | atomType atom == ErrorType]
        ++ case atom of
            CorePrepLiteral literal valueType -> verifyLiteral literal valueType
            CorePrepVariable _ _ -> []

verifyLiteral :: Core.CoreLiteral -> Type -> [Diagnostic]
verifyLiteral literal valueType =
    [problem "VXC0023" "CorePrep literal payload does not match its declared scalar type" | not matches]
    where
        matches = case literal of
            Core.CoreInteger value -> integerFits valueType value
            Core.CoreFloating spelling -> isFloatingType valueType && validFloatingSpelling spelling
            Core.CoreString _ -> valueType == stringType
            Core.CoreBoolean _ -> valueType == boolType
            Core.CoreUnit -> valueType == unitType

typeSpelling :: Type -> String
typeSpelling (NamedType (QualifiedName [Identifier name]) []) = name
typeSpelling _ = ""

integerWidths :: [(String, Bool, Int)]
integerWidths =
    [ ("char", False, 32)
    , ("byte", True, 8)
    , ("short", True, 16)
    , ("long", True, 32)
    , ("int", True, 64)
    , ("longint", True, 128)
    , ("ubyte", False, 8)
    , ("ushort", False, 16)
    , ("ulong", False, 32)
    , ("uint", False, 64)
    , ("ulongint", False, 128)
    ]

integerFits :: Type -> Integer -> Bool
integerFits valueType value = case [(signed, width) | (name, signed, width) <- integerWidths, name == typeSpelling valueType] of
    [(True, width)] -> value >= negate (2 ^ (width - 1)) && value <= 2 ^ (width - 1) - 1
    [(False, width)] -> value >= 0 && value <= 2 ^ width - 1
    _ -> False

isIntegerType :: Type -> Bool
isIntegerType valueType = any (\(name, _, _) -> name == typeSpelling valueType) integerWidths

isSignedIntegerType :: Type -> Bool
isSignedIntegerType valueType = any (\(name, signed, _) -> signed && name == typeSpelling valueType) integerWidths

isFloatingType :: Type -> Bool
isFloatingType valueType = typeSpelling valueType `elem` ["sfloat", "lfloat", "float", "double"]

isNumericType :: Type -> Bool
isNumericType valueType = isIntegerType valueType || isFloatingType valueType

validFloatingSpelling :: String -> Bool
validFloatingSpelling spelling
    | spelling `elem` ["nan", "+nan", "-nan", "inf", "+inf", "-inf"] = True
    | otherwise = case stripSign spelling of
        [] -> False
        unsignedSpelling ->
            let (mantissa, exponentPart) = break (`elem` "eE") unsignedSpelling
             in validSignificand mantissa && validExponent exponentPart
    where
        stripSign ('+' : remaining) = remaining
        stripSign ('-' : remaining) = remaining
        stripSign value = value
        validSignificand value = case break (== '.') value of
            (whole, []) -> not (null whole) && all digit whole
            (whole, _ : fraction) -> (not (null whole) || not (null fraction)) && all digit whole && all digit fraction
        validExponent [] = True
        validExponent (_ : remaining) = case stripSign remaining of
            [] -> False
            digits -> all digit digits
        digit character = character >= '0' && character <= '9'

atomType :: CorePrepAtom -> Type
atomType atom = case atom of CorePrepVariable _ valueType -> valueType; CorePrepLiteral _ valueType -> valueType

duplicateIds :: (Eq a) => String -> String -> [a] -> [Diagnostic]
duplicateIds code message values = [problem code message | length values /= length (nub values)]

problem :: String -> String -> Diagnostic
problem code message = Diagnostic CorePrepStage Error code Nothing message
